Optical process for Raman scattering in water

arturof opened the pull request:
Added a new optical process to simulate Raman scattering in water. It is a copy of the Rayleigh process with a different mean free path and a wavelength shift after the interaction based on "APPLIED OPTICS / Vol. 37, No. 15 / 20 May 1998".

The Raman process is disabled in this commit, but it can be enabled by commenting out the line added to jobOptions.mac.

Also added the ability to use the Rayleigh scattering mean free path as calculated in Geant4 internally, instead of the hardcoded values in WCSimConstructMaterials.cc from SKdetsim that are probably tuned to SK data. I use Geant4's Rayleigh mean free path when enabling the Raman process to avoid some double counting as the hardcoded values may contain a Raman component.

arturof committed:
Added new optical process for Raman scattering

The Raman process is a copy of the Rayleigh process in Geant4 with a
different mean free path and a wavelength shift after the interaction.
Created a local copy of two Geant4 enumerations to include the new Raman
process in them:
G4OpProcessSubType -> WCSimOpProcessSubType
G4OpticalProcessIndex -> WCSimOpticalProcessIndex

arturof committed:
Added Raman to the optical physics messenger

It is possible to activate/deactivate the Raman process in
jobOptions.mac. Added line in jobOptions.mac to deactivate
it because it is activated by default.

arturof committed:
Added option to use Geant4 Rayleigh mean free path

Geant4 values for Rayleigh mean free path in water can be used instead
of SKdetsim values in WCSimContructionMaterials.cc by setting rayff<=0
in tuning parameters.

arturof committed:
Added number of Raman scatters in photon history

kmtsui commented:
@arturof Thanks for the development! We do want to have Raman scattering available for a long time.

To validate your implementation, can you make some plots showing the effect from the new Raman model? For example, can you do some mono-energetic photon bomb simulation, and check the photon energy & fraction of scattered photons when being detected at the PMTs? 

And now because photon energy can change during propagation (maybe even multiple times), I think it's necessary to record the photon energy in some way. Either store the initial and final energy in `WCSimRootCherenkovHitTime`, or the sequence of energy change in `WCSimRootCherenkovHitHistory`.

kmtsui commented on src/WCSimConstructMaterials.cc:
```diff
@@ -1615,7 +1615,7 @@ void WCSimDetectorConstruction::ConstructMaterials()
    myMPT1->AddProperty("ABSLENGTH",ENERGY_water, ABSORPTION_water, NUMENTRIES_water);
    // M Fechner: new, don't let G4 compute it.
-   myMPT1->AddProperty("RAYLEIGH",ENERGY_water,RAYLEIGH_water,NUMENTRIES_water);
+   if (RAYFF>0) myMPT1->AddProperty("RAYLEIGH",ENERGY_water,RAYLEIGH_water,NUMENTRIES_water);
```
I think this is related to your comment of `Also added the ability to use the Rayleigh scattering mean free path as calculated in Geant4 internally, instead of the hardcoded values in WCSimConstructMaterials.cc from SKdetsim that are probably tuned to SK data.`

Better to add explain in `macros/tuning_parameters.mac` on what would happen when `/WCSim/tuning/rayff 0` is used

arturof replied:
Yes, it is related to that. I will add an explanation in `macros/tuning_parameters.mac`.

kmtsui commented on src/WCSimConstructMaterials.cc:
```diff
@@ -1678,7 +1678,7 @@ void WCSimDetectorConstruction::ConstructMaterials()
    SilGelPropTable->AddProperty("RINDEX", ENERGY_water, RINDEX_SilGel, NUMENTRIES_water);
    SilGelPropTable->AddProperty("ABSLENGTH",ENERGY_SilGel, ABSORPTION_SilGel, 18); //ToDo: get measurement of optical properties of the optical gel. From slides: better than 40cm above 350nm.
-   SilGelPropTable->AddProperty("RAYLEIGH",ENERGY_water,RAYLEIGH_water,NUMENTRIES_water); //ToDo: get actual Rayleigh scattering in gel
+   if (RAYFF>0) SilGelPropTable->AddProperty("RAYLEIGH",ENERGY_water,RAYLEIGH_water,NUMENTRIES_water); //ToDo: get actual Rayleigh scattering in gel
```
This is for silicon gel. Do not couple with water implementation.

arturof replied:
I am also wondering why the Rayleigh mean free path for water is used for silicon gel. Somebody probably concluded that it is better than not having Rayleigh scattering in silicon gel at all.

The problem is that when RAYFF<=0, the photon entering the silicon gel will Rayleigh scatter all the time or generate an error. We could remove that line and put it back when the correct Rayleigh mean free path for silicon gel is available.

kmtsui commented on src/WCSimOpRaman.cc:
```diff
@@ -0,0 +1,289 @@
+  // If the material doesn't have a RINDEX property vector then return
+  G4MaterialPropertyVector* rIndex = materialProperties->GetProperty("RINDEX");
+  if ( rIndex == NULL ) return NULL;
+
```
Does that mean Raman scattering is enabled for all materials with `RINDEX` table? Is this intended? Or should it be restricted to Water/GdWater?

arturof replied:
Yes, it is currently enabled for all materials with `RINDEX` table. I overlooked that by mistake. I will restrict it to Water and GdWater.

kmtsui commented on src/WCSimOpRaman.cc:
```diff
@@ -0,0 +1,289 @@
+        const G4DynamicParticle* aParticle = aTrack.GetDynamicParticle();
+
+        if (verboseLevel>0) {
+                G4cout << "Scattering Photon!" << G4endl;
```
Just for more clear verbose
```
G4cout << "Raman scattering Photon!" << G4endl;
```

arturof replied:
Ok, that makes sense.

kmtsui commented on src/WCSimOpRaman.cc:
```diff
@@ -0,0 +1,289 @@
+  {
+     const G4double energy = rIndex->Energy( uRIndex );
+     const G4double lambda = h_Planck * c_light / energy;
+     const G4double meanFreePath = (3703.7037*m) * pow( lambda / (488*nm) , 5.3);
```
Can you plot mean free path/cross section for visualization?

arturof replied:
The attached plot compares the mean free path in water of the new Raman process with other optical processes. The absorption and Rayleigh labels refer to WCSim default values. G4Rayleigh to the Rayleigh values calculated by Geant4. 
![wcsim_water_mfp](https://github.com/user-attachments/assets/c15d5e0d-0487-41df-9c94-f3cca26ae261)

arturof committed:
Added explanation of changes in mac files

Explanation for rayff usage in tuning_parameters.mac and the line
to activate/deactivate optical processes in jobOptions.mac

arturof committed:
More clear verbose

arturof committed:
Restrict Raman process to water and Gd doped water

arturof committed:
Added start and end photon energy to output file

arturof committed:
Merge branch 'develop' into raman

arturof commented:
To validate the Raman model, I extracted the total attenuation length by fitting the true interaction position distribution in a simulation of 10000 295nm photons starting at the barrel wall and pointing along the cylinder diameter. The expected value is 1/(1/Absorption + 1/Rayleigh + 1/Raman) = 1/(1/571m + 1/41.1m + 1/257m) = 33.4m and the fit result is close to it. 

<img width="696" height="472" alt="WCSim_HK_pencil_295nm_RamanOn_ntuple_all_pos" src="https://github.com/user-attachments/assets/cb926be9-acd1-49d1-9ecb-6962b6c22ad3" />

I also checked the final photon wavelength at the PMTs. The Raman model shifts the initial wavelength according to 1/lambda_final = 1/lambda_initial - 3400cm^-1 (it increases by ~30nm at 295nm and ~40nm at 320nm). The first peak in the plot contains photons that didn't scatter or did Rayleigh scattering only. The second peak contains photons that did 1 Raman interaction and the third peak 2 Raman interactions. The first peak is suppressed due to the low PMT QE at 295nm.

<img width="696" height="472" alt="WCSim_HK_pencil_295nm_RamanOn_allPMT_histos_Lambda10_scat" src="https://github.com/user-attachments/assets/6a5048ae-527f-4dc8-9b52-269e041b1031" />

arturof commented:
I am storing the initial and final photon energy in `WCSimRootCherenkovHitTime` and only the total number of Raman interactions in `WCSimRootCherenkovHitHistory` for now. Do I need to increase the version number of these classes in include/WCSimRootEvent.hh?

I plan to store the photon position, energy and interaction type of each step in `WCSimRootCherenkovHitHistory`. Is this reasonable? Is `WCSimRootCherenkovHitHistory` the correct place to do it? In addition to keep track of the energy changes due to the Raman interaction, I need this information to be able to re-weight the absorption, Rayleigh and Raman interaction probabilities.

kmtsui commented on src/WCSimWCSD.cc:
```diff
@@ -412,6 +424,8 @@ G4bool WCSimWCSD::ProcessHits_boundary(G4Step* aStep, G4TouchableHistory*)
         (*hitsCollection)[PMTHitMap[replicaNumber]-1]->AddParentID(parentSavedTrackID);
         (*hitsCollection)[PMTHitMap[replicaNumber]-1]->AddPhotonStartTime(photonStartTime);
+        (*hitsCollection)[PMTHitMap[replicaNumber]-1]->AddPhotonStartEnergy(photonStartEnergy);
+        (*hitsCollection)[PMTHitMap[replicaNumber]-1]->AddPhotonStartEnergy(photonEndEnergy);
```
Bug here

kmtsui replied:
seems these two bugs cause segmentation fault in WCTE tests.

arturof replied:
Thanks for catching that. Somehow it didn't crash when I run it using the HK geometry. I fixed both bugs.

kmtsui commented on src/WCSimWCSD.cc:
```diff
@@ -424,6 +438,8 @@ G4bool WCSimWCSD::ProcessHits_boundary(G4Step* aStep, G4TouchableHistory*)
         (*hitsCollection)[PMTHitMap[replicaNumber]-1]->AddParentID(parentSavedTrackID);
         (*hitsCollection)[PMTHitMap[replicaNumber]-1]->AddPhotonStartTime(photonStartTime);
+        (*hitsCollection)[PMTHitMap[replicaNumber]-1]->AddPhotonStartEnergy(photonStartEnergy);
+        (*hitsCollection)[PMTHitMap[replicaNumber]-1]->AddPhotonStartEnergy(photonEndEnergy);
```
Bug here

kmtsui commented:
> I am storing the initial and final photon energy in `WCSimRootCherenkovHitTime` and only the total number of Raman interactions in `WCSimRootCherenkovHitHistory` for now. Do I need to increase the version number of these classes in include/WCSimRootEvent.hh?

Yes it is.

> I plan to store the photon position, energy and interaction type of each step in `WCSimRootCherenkovHitHistory`. Is this reasonable? Is `WCSimRootCherenkovHitHistory` the correct place to do it? In addition to keep track of the energy changes due to the Raman interaction, I need this information to be able to re-weight the absorption, Rayleigh and Raman interaction probabilities.

I agree.

arturof committed:
Fixed bug setting photon end energy in WCSimWCSD

tdealtry commented on macros/tuning_parameters.mac:
```diff
@@ -1,4 +1,6 @@
 # Set tuning parameters
+# Set rayff>0 to scale the Rayleigh mean free path in water hard-coded in WCSimContructMaterials.cc
+# Set rayff<=0 to use the Rayleigh mean free path in water calculated by Geant4
```
```suggestion
# Set rayff>0 to scale the Rayleigh mean free path in water hard-coded in WCSimContructMaterials.cc, based on SK data
# Set rayff<=0 to use the Rayleigh mean free path in water calculated by Geant4
# If activating Raman scattering, using the Geant4 value is the current recommendation (the SK measurement of Rayleigh scattering may have a Raman effect)
```

tdealtry commented on include/WCSimOpRaman.hh:
```diff
@@ -0,0 +1,188 @@
+// * conditions of the Geant4 Software License,  included in the file *
+// * LICENSE and available at  http://cern.ch/geant4/license .  These *
+// * include a list of copyright holders.                             *
+// *                                                                  *
+// * Neither the authors of this software system, nor their employing *
+// * institutes,nor the agencies providing financial support for this *
+// * work  make  any representation or  warranty, express or implied, *
+// * regarding  this  software system or assume any liability for its *
+// * use.  Please see the license in the file  LICENSE  and URL above *
+// * for the full disclaimer and the limitation of liability.         *
+// *                                                                  *
+// * This  code  implementation is the result of  the  scientific and *
+// * technical work of the GEANT4 collaboration.                      *
+// * By using,  copying,  modifying or  distributing the software (or *
+// * any work based  on the software)  you  agree  to acknowledge its *
+// * use  in  resulting  scientific  publications,  and indicate your *
+// * acceptance of all terms of the Geant4 Software license.          *
+// ********************************************************************
+//
+//
+// $Id: G4OpRayleigh.hh 84596 2014-10-17 07:34:41Z gcosmo $
+//
+// 
+////////////////////////////////////////////////////////////////////////
+// Optical Photon Rayleigh Scattering Class Definition
+////////////////////////////////////////////////////////////////////////
+//
+// File:        G4OpRayleigh.hh
+// Description: Discrete Process -- Rayleigh scattering of optical photons 
+// Version:     1.0
+// Created:     1996-05-31
+// Author:      Juliet Armstrong
+// Updated:     2014-08-20 allow for more material types
+//              2005-07-28 add G4ProcessType to constructor
+//              1999-10-29 add method and class descriptors
+//              1997-04-09 by Peter Gumplinger
+//              > new physics/tracking scheme
+// mail:        gum@triumf.ca
+//
+////////////////////////////////////////////////////////////////////////
```
From what I can tell, you just copied the Rayleigh class, and renamed the class / rewrote the source
This was instead of setting up a brand new class that inherits from `G4VDiscreteProcess` directly
Both would give the same results (one is a bit quicker!) so I don't think this header is required
```suggestion
```

tdealtry commented on include/WCSimOpRaman.hh:
```diff
@@ -0,0 +1,188 @@
+        G4double GetMeanFreePath(const G4Track& aTrack,
+				 G4double ,
+                                 G4ForceCondition* );
+        // Returns the mean free path for Rayleigh scattering in water.
```
```suggestion
        // Returns the mean free path for Raman scattering in water.
```

tdealtry commented on include/WCSimOpRaman.hh:
```diff
@@ -0,0 +1,188 @@
+
+        G4VParticleChange* PostStepDoIt(const G4Track& aTrack,
+                                       const G4Step&  aStep);
+        // This is the method implementing Rayleigh scattering.
```
```suggestion
        // This is the method implementing Raman scattering.
```

tdealtry commented on include/WCSimOpRaman.hh:
```diff
@@ -0,0 +1,188 @@
+}
+
+
+#endif /* G4OpRayleigh_h */
```
```suggestion
#endif /* WCSimOpRaman_h */
```

tdealtry commented on include/WCSimOpRaman.hh:
```diff
@@ -0,0 +1,188 @@
+        // Methods
+        ////////////
+
+        G4bool IsApplicable(const G4ParticleDefinition& aParticleType);
+        // Returns true -> 'is applicable' only for an optical photon.
```
```suggestion
        /// Returns true -> 'is applicable' only for an optical photon.
        G4bool IsApplicable(const G4ParticleDefinition& aParticleType);
```
Since we use doxygen, and I guess they use something bespoke, it'd be good to do this. Essentially
- move the comment line(s) above, rather than below the function
- add an extra `/` to make it a triple `///`

This is a general comment for this (and other) files. But if it's a massive pain, then don't worry

tdealtry commented on src/WCSimOpRaman.cc:
```diff
@@ -0,0 +1,293 @@
+//              > new physics/tracking scheme
+// mail:        gum@triumf.ca
+//
+////////////////////////////////////////////////////////////////////////
```
Again, I think you can delete this
```suggestion
```

tdealtry reviewed (changes requested):
Thanks @arturof Looks great! Thanks for bringing in a new feature like this!

Please increment `ClassDef`
- [Here](https://github.com/WCSim/WCSim/blob/develop/include/WCSimRootEvent.hh#L198)
- [And here](https://github.com/WCSim/WCSim/blob/develop/include/WCSimRootEvent.hh#L226)

Is it required to add new Raman enums [here](https://github.com/tdealtry/WCSim/blob/fc6b3a8cb8104157aad52f437e4536d216c57510/include/WCSimEnumerations.hh#L60)? e.g. `kOpRaman` (or `fOpRaman` - I'm confused given WCSimOpProcessSubType)

arturof committed:
Update macros/tuning_parameters.mac

Co-authored-by: Tom Dealtry <tom.deal135@gmail.com>

arturof committed:
Update include/WCSimOpRaman.hh

Co-authored-by: Tom Dealtry <tom.deal135@gmail.com>

arturof committed:
Update include/WCSimOpRaman.hh

Replace "Rayleigh" with "Raman" in comments

Co-authored-by: Tom Dealtry <tom.deal135@gmail.com>

arturof committed:
Update include/WCSimOpRaman.hh

Replace "Rayleigh" with "Raman" in comments

Co-authored-by: Tom Dealtry <tom.deal135@gmail.com>

arturof committed:
Update include/WCSimOpRaman.hh

Replace "Rayleigh" with "Raman" in comments

Co-authored-by: Tom Dealtry <tom.deal135@gmail.com>

arturof committed:
Update src/WCSimOpRaman.cc

Remove G4OpRayleigh header

Co-authored-by: Tom Dealtry <tom.deal135@gmail.com>

arturof committed:
Update include/WCSimOpRaman.hh

Co-authored-by: Tom Dealtry <tom.deal135@gmail.com>

arturof committed:
Merge branch 'develop' into raman

arturof committed:
Make WCSimOpRaman.hh comments doxygen compatible

arturof committed:
Add Raman to WCSim enumerations for completeness

tdealtry commented:
Thanks for implementing my suggestions @arturof 

Think there's just the ClassDef request & enum question, then this can be merged. Thanks again for your work on this!

arturof committed:
Add scattering and reflection positions to photon history

arturof committed:
Increase version of WCSimRootCherenkovHitHistory and WCSimRootCherenkovHitTime

arturof commented:
Thanks for your input @tdealtry

I don't think it is required to have the Raman process in the EProcessType enumeration. It seems that it is used only to save the process that created a photon that created a hit, but the Raman process doesn't create photons. However, I added the Raman process anyway because the other optical processes (Rayleigh, Mie, absorption and boundary which don't create photons either) are included.

I incremented ClassDef as requested, but also added an additional feature that I mentioned before and was planing to include in this PR: I added the photon scattering and reflection positions to `WCSimRootCherenkovHitHistory` to be able to re-weight the mean free path of the Raman process and the other optical processes. I could open another PR for it if needed.

tdealtry reviewed (approved):
Thanks @arturof
